src/utils/clusterUtils.ts

```diff
@@ -0,0 +1,417 @@
+export function getCpusTotalNodePools(nodePools: INodePool[]) {
+  const provider = guessProviderFromNodePools(nodePools);
+  if (provider === null) {
+    return 0;
+  }
+
+  const instanceTypes = getInstanceTypesForProvider(provider);
+  if (instanceTypes === null) {
+    return 0;
+  }
+
+  /**
+   * Here we are returning (and accumulating) for each node pool the number
+   * of CPUs each instance has multiplied by the number of nodes the node pool has.
+   *
+   * TODO: When working with Spot Instances a node pool could have different types
+   * of instances, and this method will have to be modified.
+   */
+  const totalCPUs = nodePools.reduce((accumulator, nodePool) => {
+    let instanceTypeCPUs = 0;
+    let currInstanceType = '';
+
+    switch (provider) {
+      case Providers.AWS:
+        currInstanceType = nodePool.node_spec.aws?.instance_type ?? '';
+        instanceTypeCPUs =
+          (instanceTypes as Record<string, IRawAWSInstanceType>)[
+            currInstanceType
+          ]?.cpu_cores ?? 0;
+
+        break;
+
+      case Providers.AZURE:
+        currInstanceType = nodePool.node_spec.azure?.vm_size ?? '';
+        instanceTypeCPUs =
+          (instanceTypes as Record<string, IRawAzureInstanceType>)[
+            currInstanceType
+          ]?.numberOfCores ?? 0;
+
+        break;
+    }
+
+    instanceTypeCPUs = instanceTypeCPUs * nodePool.status.nodes_ready;
+
+    return accumulator + instanceTypeCPUs;
+  }, 0);
+
+  return totalCPUs;
+}
+
+/**
+ * Determine which features that a cluster supports.
+ * @param state - The app's global state.
+ */
+export const computeCapabilities = (state: IState) => (
+  releaseVersion: string,
+  provider: PropertiesOf<typeof Providers>
+): IClusterCapabilities => {
+  let hasOptionalIngress = false;
+  let supportsHAMasters = false;
+
+  const minHAMastersVersion = getMinHAMastersVersion(state);
+
+  switch (provider) {
+    case Providers.AWS:
+      hasOptionalIngress = cmp(releaseVersion, '10.0.99') === 1;
+      supportsHAMasters = cmp(releaseVersion, minHAMastersVersion) >= 0;
+
+      break;
+
+    case Providers.AZURE:
+      hasOptionalIngress = cmp(releaseVersion, '12.0.0') >= 0;
+
+      break;
+
+    case Providers.KVM:
+      hasOptionalIngress = cmp(releaseVersion, '12.2.0') >= 0;
+
+      break;
+  }
+
+  return {
+    hasOptionalIngress,
+    supportsHAMasters,
+  };
+};
+
+export function filterLabels(labels?: IClusterLabelMap) {
+  if (!labels) {
+    return undefined;
+  }
+
+  const filteredLabels: IClusterLabelMap = {};
+
+  for (const key of Object.keys(labels)) {
+    if (validateLabelKey(key).isValid) {
+      filteredLabels[key] = labels[key];
+    }
+  }
+
+  return filteredLabels;
+}
+
+/**
+ * Get a cluster's latest condition.
+ * @param cluster
+ */
+export function getClusterLatestCondition(
+  cluster: V4.ICluster | V5.ICluster
+): string | undefined {
+  if (
+    'conditions' in cluster &&
+    cluster.conditions &&
+    cluster.conditions.length > 0
```

Comment:
Not as pretty as the way we used optional chaining before, but that way confuses the compiler too much, and it gives a false error.